lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Validate.hs

```diff
@@ -0,0 +1,93 @@
+module LambdaBuffers.Compiler.TypeClass.Validate (
+  -- structural rules
+  mkStructuralRules,
+  -- pattern vars (for tests)
+  _X,
+  _XS,
+  _VARS,
+  _NAME,
+  _L,
+  _BODY,
+  -- main function
+  checkDerive,
+) where
+
+import Data.Set qualified as S
+
+import Control.Monad.Except (throwError)
+import LambdaBuffers.Compiler.ProtoCompat.Types qualified as P (
+  ModuleName,
+ )
+import LambdaBuffers.Compiler.TypeClass.Pat (Exp (DecE), Literal (Opaque), Pat (ConsP, LabelP, LitP, NilP, ProdP, RecP, SumP, VarP), getLocalRefE)
+import LambdaBuffers.Compiler.TypeClass.Rules (
+  Class,
+  Constraint (C),
+  Rule ((:<=)),
+ )
+import LambdaBuffers.Compiler.TypeClass.Solve (Overlap, inst, solve)
+import LambdaBuffers.Compiler.TypeClass.Utils (
+  BasicConditionViolation (OverlapDetected),
+  ModuleBuilder (mbInstances, mbScope, mbTyDefs),
+  TypeClassError (BadInstance, LocalTyRefNotFound, MalformedTyDef),
+  lookupOr,
+ )
+
+-- hardcoded PATTERN variables, easier to read than (VarP "blah") everywhere
+_X, _XS, _VARS, _NAME, _L, _BODY :: Pat
+_X = VarP "x"
+_XS = VarP "xs"
+_VARS = VarP "vars"
+_NAME = VarP "name"
+_L = VarP "l"
+_BODY = VarP "body"
+
+-- Create a set of structural rules for a given class
+mkStructuralRules :: Class -> [Rule Pat]
+mkStructuralRules c =
+  [ C c NilP :<= []
+  , C c (ConsP _X _XS) :<= [C c _X, C c _XS]
+  , C c (LabelP _L _X) :<= [C c _X]
+  , C c (RecP _XS) :<= [C c _XS]
+  , C c (ProdP _XS) :<= [C c _XS]
+  , C c (SumP _XS) :<= [C c _XS]
+  , C c (LitP Opaque) :<= []
+  ]
+
+-- utility
+splitInstance :: Rule a -> (Constraint a, [Constraint a])
+splitInstance (C c t :<= is) = (C c t, is)
+
+constraintClass :: Constraint a -> Class
+constraintClass (C c _) = c
+
+-- NOTE: Practically this enforces the "must define instances where types are defined"
+--       half of Haskell's orphan instances rule. We could relax that in various ways
+--       but it would require reworking a lot of the utilities above.
+checkDerive :: P.ModuleName -> ModuleBuilder -> Rule Pat -> Either TypeClassError [Constraint Exp]
+checkDerive mn mb i = concat <$> (traverse solveRef =<< catchOverlap (solve assumptions c))
+  where
+    catchOverlap :: Either Overlap a -> Either TypeClassError a
+    catchOverlap = either (Left . BadInstance . OverlapDetected) pure
+
+    (c', cs') = splitInstance i
+    c = fmap inst c'
+
+    localInstances = mbInstances mb
+
+    localTyDefs = mbTyDefs mb
+
+    inScopeInstances = mbScope mb
+
+    solveRef :: Constraint Exp -> Either TypeClassError [Constraint Exp]
+    solveRef (C cx ty) = case getLocalRefE ty of
+      Just t -> do
+        lookupOr t localTyDefs (LocalTyRefNotFound t mn) >>= \case
+          DecE _ _ body -> catchOverlap $ solve assumptions (C cx body)
+          xp -> throwError $ MalformedTyDef mn xp
+      _ -> pure [C cx ty]
```

Comment:
I guess this is it for TyRef user specifiable rule.

I don't understand why this is rammed here, this should just be added to the rule set along with other rules (structural plus ad hoc instances):

```C c (LitP RefP "Int") :<= [C c OpaqueE ]```

Reply:
I'll try to explain this next week. It only makes sense as a rule if we can express lookups directly as rules. I think?